qiskit/utils/deprecation.py

```diff
@@ -89,27 +102,96 @@ def wrapper(*args, **kwargs):
-                warnings.warn(
-                    f"{func_name} keyword argument {old_arg} is deprecated and "
-                    f"replaced with {new_arg}.",
-                    category=category,
-                    stacklevel=3,
-                )
-
-                kwargs[new_arg] = kwargs.pop(old_arg)
+        if old_arg not in kwargs:
+            continue
+        if new_arg in kwargs:
+            raise TypeError(f"{func_name} received both {new_arg} and {old_arg} (deprecated).")
+        warnings.warn(old_kwarg_to_msg[old_arg], category=category, stacklevel=3)
+        if new_arg is not None:
+            kwargs[new_arg] = kwargs.pop(old_arg)
+
+
+def _add_deprecation_to_docstring(
+    func: Callable, msg: str, *, since: Optional[str], pending: bool
+) -> None:
+    if "\n" in msg:
+        raise ValueError(
+            "Deprecation messages cannot contain new lines (`\\n`), but the deprecation for "
+            f'{func.__qualname__} had them. Usually this happens when using `"""` multiline '
+            f"strings; instead, use string concatenation.\n\n"
+            "This is a simplification to facilitate deprecation messages being added to our "
+            "documentation. If you have a compelling reason to need "
+            "new lines, feel free to improve this function or open a request at "
+            "https://github.com/Qiskit/qiskit-terra/issues."
+        )
+
+    if since is None:
+        version_str = "unknown"
+    else:
+        version_str = f"{since}_pending" if pending else since
+
+    # We insert deprecations in-between the description and the meta sections. The below is from
+    # https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#docstring-sections.
+    # We only include what is used by Qiskit projects for functions/methods to avoid wasting
+    # computation.
+    meta_lines = {
+        "Args:",
+        "Arguments:",
+        "Example:",
+        "Examples:",
+        "Note:",
+        "Notes:",
+        "Parameters:",
+        "Return:",
+        "Returns:",
+        "Raises:",
+        "References:",
+        "Yield:",
+        "Yields:",
+    }
+
+    indent = ""
+    meta_index = None
+    if func.__doc__:
+        original_lines = func.__doc__.splitlines()
+        content_encountered = False
+        for i, line in enumerate(original_lines):
+            stripped = line.strip()
+
+            # Determine the indent based on the first line with content. But, we don't consider the
+            # first line, which corresponds to the format """Docstring.""", as it does not properly
+            # capture the indentation of lines beneath it.
+            if not content_encountered and i != 0 and stripped:
+                num_spaces = len(line) - len(line.lstrip())
+                indent = " " * num_spaces
+                content_encountered = True
+
+            if stripped in meta_lines:
+                meta_index = i
+                assert content_encountered is True
```

Comment:
I'd maybe turn this into an exception with a message saying "docstrings need some text before the first section heading" or something, just as a hint to developers.  It's fairly common to write the class summary in the class docstring and then start the `__init__` method's with `Args:` directly - that won't cause directly issues here I think because we configure Sphinx to merge those two docstrings, but the point is that it _is_ a thing somebody might occasionally do elsewhere by mistake.

Reply:
So, this algorithm will handle that, as shown by the tests. This is truly an assertion meant to catch if my assumptions in the algorithm are wrong.

I'll make it a proper AssertionError with text so this is more clear.

Reply:
The reason there could be an error is because the conditions between the two `if`s aren't exactly the same.  If you make a docstring that has no blank line at the start and has a first line `Args:`, it'll raise this assertion error.  It's unlikely to be used as a docstring given our lint rules, but it's an uncomfortable coupling between lint rules and this function, especially because your intent is for this function to be in the public API - we don't enforce our lint rules on others.

I'd vote in favour of refactoring the logic to not have the assertion - even if the very very first line is a section heading, there's still a sensible place to put the deprecations (before it).

Reply:
![Screenshot 2023-03-04 at 00 27 31](https://user-images.githubusercontent.com/5968590/222860065-1b179798-5b24-43d1-911b-392e9bdd483d.png)

Reply:
Thanks for catching this legit bug! As discussed over Slack, supporting this style is quite tricky to get indentation right. So, I turned this into a proper `ValueError` and suggest an easy fix for authors to add one new line:

> ValueError: add_deprecation_to_docstring cannot currently handle when a Napoleon metadata line like 'Args' is the very first line of docstring, e.g. `"""Args:`. So, it cannot process Grover.quantum_instance. Instead, move the metadata line to the second line, e.g.:
>
> """
> Args: